22.0.0.10-beta draft (#2498)
OpenLiberty/blogs: posts/2022-09-01-22.0.0.10-beta.adoc

* 425: Virtual Threads (Preview)
* 426: Vector API (Fourth Incubator)
* 427: Pattern Matching for switch (Third Preview)
* 428: Structured Concurrency (Incubator)

By taking advantage of trying out the new changes in Java 19 now, you will get more time to review your applications, microservices and runtime environments which means you will be a step ahead when it becomes generally available.

To give this a try, link:https://jdk.java.net/19/[download the early access release of Java 19], download and install the link:https://openliberty.io/downloads/#runtime_betas[22.0.0.10-beta] version of Open Liberty, edit your Liberty server's link:https://openliberty.io/docs/22.0.0.10/reference/config/server-configuration-overview.html#server-env[server.env] file to point JAVA_HOME to your Java 19 installation and start testing!

For more information on Java 19, visit the Java 19 link:https://jdk.java.net/19/release-notes[release notes page], link:https://download.java.net/java/early_access/jdk19/docs/api/[API Javadoc page] or link:https://openliberty.io/docs[download page].
For more information on Open Liberty, please visit our documentation page.

> As work is made towards full Java 19 support, there may functionality that might not be 100% ready yet but is in progress to be complete soon!

Reviewer: ```suggestion
As work is made towards full Java 19 support, there may be functionality that might not be 100% ready yet but is in progress to be complete soon!

```

// https://github.com/OpenLiberty/open-liberty/issues/22092
[#xml]
== XML Binding 4.0

> Jakarta XML Binding 4.0 is an API for automating the mapping between XML documents and Java objects. Support provides the updated XML Binding 4.0 API, it's Reference Implementation, and introduces the xmlBinding-4.0 feature to Open Liberty.

Reviewer 2: "for automating the mapping between XML documents and Java objects when writing XML documents in Java" or something like that? Can we give a brief sentence about when you'd use this?

Reviewer: @neuwerk could you at least include something about what changed in 4.0 compared to 3.0?  I think that would help for sure and I think Laura would like an example of how it is used or what you can do now that you couldn't do before.

Reviewer 2: no changes

> Jakarta XML Binding 4.0 is an API for automating the mapping between XML documents and Java objects. Support provides the updated XML Binding 4.0 API, it's Reference Implementation, and introduces the xmlBinding-4.0 feature to Open Liberty.

Reviewer: ```suggestion
Jakarta XML Binding 4.0 is an API for automating the mapping between XML documents and Java objects. Support provides the updated XML Binding 4.0 API, its Reference Implementation, and introduces the xmlBinding-4.0 feature to Open Liberty.

```

> The xmlBinding-4.0 feature provides the Jakarata XML Binding 4.0 tools for marshalling and unmarshing XML to Java: xjc and schemagen. Both command line tools preform identically to the Jakarata EE 9 tooling.

Reviewer: ```suggestion
The xmlBinding-4.0 feature provides the Jakarta XML Binding 4.0 tools for marshaling and unmarshaling XML to Java: xjc and schemagen. Both command line tools perform identically to the Jakarta EE 9 tooling.

```

To enable the new Jakarta XML Binding 4.0 feature, add the xmlBinding-4.0 feature to your `server.xml`. Here's an example of what the feature configuration looks like:

[source,xml]
----
    <featureManager>
          <feature>xmlBinding-4.0</feature>
    </featureManager>  
----

For more information see the link:https://jakarta.ee/specifications/xml-binding/[Jakarta XML Binding Specification]

// https://github.com/OpenLiberty/open-liberty/issues/21974
[#bell]

> == BELL SPI Visibility and BELL Properties

Reviewer 2: This section seems like a whole blog post embedded into this release blog post. Should it be broken out into a separate post and then just have a summary in this beta release blog post that links to it instead?

Reviewer: It definitely could be shortened by removing the repeated statements.

The link:https://openliberty.io/docs/latest/reference/feature/bells-1.0.html[Basic Extensions using Liberty Libraries (BELL) 1.0] feature enables shared libraries to provide implementations of Liberty API interfaces using Java link:https://docs.oracle.com/javase/9/docs/api/java/util/ServiceLoader.html[ServiceLoader] configuration files.

> 22.0.0.10-beta introduces two capabilities for BELL services: SPI visibility and properties configuration and Injection.

Reviewer 2: ```suggestion
22.0.0.10-beta introduces two capabilities for BELL services: SPI visibility, and properties configuration and injection.

```

[source, xml]
----
        <library id=”ServletExtLib">
            <fileset dir="${server.output.dir}/sharedLib" includes="ServletExtLib.jar" />
        </library>

        <bell libraryRef=”ServletExtLib”
              service="javax.servlet.ServletContainerInitializer"
              spiVisibility=”true">
           <properties servletName=“RestSciServlet” servletMapping="/rest/*" />
        </bell>
----

In order to receive properties at runtime, modify your service implementation classes by defining either a public constructor or a public method named `updateBell`, each having a single parameter of type `java.util.Map<String,String>`. The `updateBell` method is intended to support non-destructive updates and must tolerate multiple calls on the same service instance.

The example below defines method `updateBell` within service implementation class `RestSci`.

[source, java]
----
        package my.servlet.extensions;
        ...    
        public class RestSci implements javax.servlet.ServletContainerInitializer {

>            public RestfSci() {}

Reviewer: ```suggestion
           public RestSci() {}

```

----

When the server registers the BELL service with the OSGi framework it triggers the REST Handler framework to register the service as a listener for URL sub-root `/bellEP`. The /bellEP endpoint is now accessible.

You can now make calls to `https://<host>:8020/ibm/api/bellEP`.

The REST handler framework references the BELL service to serve the `/bellEP` endpoint. During the first reference the server creates an instance of class BellEndpoint and injects an unmodifiable map with key/value pair `"hello"/"WORLD"`. The server injects BELL properties at service creation and after BELL configuration updates.

You can find out more about this specific update from the following links:

* link:https://openliberty.io/docs/latest/reference/feature/bells-1.0.html[Open Liberty BELL 1.0 Feature Doc]
* link:https://docs.oracle.com/javase/9/docs/api/java/util/ServiceLoader.html[Java ServiceLoader]
* link:https://openliberty.io/docs/latest/reference/config/bell.html[Open Liberty BELL Server Configuration Doc]

// https://github.com/OpenLiberty/open-liberty/issues/21847
[#microprofile]

> == MicroProfile Telemetry 1.0

Reviewer 2: Could this be higher up the post (as it's MicroProfile)?

> link:https://opentelemetry.io/[[OpenTelemetry] is a set of APIs, SDKs, tooling and integrations that are designed for the creation and management of telemetry data such as traces, metrics, and logs.

Reviewer 2: I think this needs another sentence describing the context in which telemetry data might be used. eg is this for tracing activity between microservices? Or within a single service? Can we give a brief example scenario to really make it clear what this is for and why you should find it useful?

> link:https://opentelemetry.io/[[OpenTelemetry] is a set of APIs, SDKs, tooling and integrations that are designed for the creation and management of telemetry data such as traces, metrics, and logs.

Reviewer: ```suggestion
link:https://opentelemetry.io/[OpenTelemetry] is a set of APIs, SDKs, tooling and integrations that are designed for the creation and management of telemetry data such as traces, metrics, and logs.

```

[source, xml]
----
<featureManager>
  <feature>mpTelemetry-1.0</feature>
</featureManager>
----

You can view the telemetry data from your application in link:https://www.jaegertracing.io/docs/1.37/monitoring/#traces[Jaeger] and  link:https://zipkin.io/pages/quickstart.htmlp[Zipkin]. To export your traces to one of these exporters, configure the exporter with system properties, e.g. `-Dotel.traces.exporter=jaeger` or environment variables, e.g. `OTEL_TRACES_EXPORTER=jaeger`.  You can also configure the exporter endpoints using `OTEL_EXPORTER_JAEGER_ENDPOINT` and `OTEL_EXPORTER_ZIPKIN_ENDPOINT`.  To specify the name of the service, use `OTEL_SERVICE_NAME`.
   
For more information about MicroProfile Telemetry, see the following links:

* link:https://github.com/eclipse/microprofile-telemetry[microprofile-telemetry]
* link:https://github.com/open-telemetry/opentelemetry-specification/blob/v1.11.0/specification/trace/api.md[opentelemetry-specification]
* link:https://opentelemetry.io[opentelemetry.io]

// https://github.com/OpenLiberty/open-liberty/issues/20209
[#security]

> == Distributed Security Caches

Reviewer 2: Again, this section is really long for a release blog post. Should it be broken out into a separate blog post?

Reviewer 2: Removed

>    - A sentence or two that introduces the update to someone new to the general technology/concept.

Reviewer: This left in some of the extra stuff that shouldn't be in here.  Not sure if this feature needs to be in this blog post since it has been in since 22.0.0.4-beta:  https://openliberty.io/blog/2022/03/17/graphql-22004-beta.html#security_caches

Reviewer 2: We've removed it now.

   
 > Distributed security cache support has been introduced so that multiple Liberty servers can share caches via a JCache provider.

   - What was the problem before and how does your update make their life better? (Why should they care?)

> > Prior to this feature, the authentication (subject) and logged out cookie caches were restricted to be local and in-memory. Multiple servers were unable to benefit from their peers' caches and each server started with a cold cache.

Reviewer 2: Can we give a more concrete example of why servers starting with a cold cache is a problem that needed solving? From this I don't get a sense of why I should care about distributed security caches. Yet, I know they're useful because we had this nice blog post in the past about persistence using JCache: https://openliberty.io/blog/2018/03/22/distributed-in-memory-session-caching.html Having a concrete example like a shopping trolley makes it clearer what the benefit is. How would starting with a cold cache be a problem in that example (if relevant; if not, is there a better example for this situation?)?

Reviewer 2: Removed

     -->
    <cacheManager id="CacheManager" uri="uri://someuri">
        <properties prop1="value1" prop2="value2" />

        <cachingProvider libraryRef="JCacheProviderLib" />
    </cacheManager>

    <!-- 
        Configure the JCache cache instances.
     -->
    <cache id="AuthCache" name="AuthCache" cacheManagerRef="CacheManager" />
    <cache id="LoggedOutCookieCache" name="LoggedOutCookieCache" cacheManagerRef="CacheManager" />

    <!-- 
        Configured the authentication cache, logged-out cookie cache and HTTP session cache.
     -->
    <authCache cacheRef="AuthCache" />
    <webAppSecurity loggedoutCookieCacheRef="LoggedOutCookieCache" ... />
    <httpSessionCache cacheManagerRef="CacheManager" ... />
```
   
   - Where can they find out more about this specific update (eg Open Liberty docs, Javadoc) and/or the wider technology?

- https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-configuring-authentication-cache-in
- https://www.ibm.com/docs/en/was-liberty/base?topic=manually-configuring-liberty-session-persistence-jcache

To enable the new beta features in your app, add them to your `server.xml`:

[source, xml]
----

----

[#run]
=== Try it now 

> To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE XX, Java SE 17, Java SE 11, and Java SE 8.
> // // // // // // // //
> // In the preceding section:
> // Check if a new non-LTS Java SE version is supported that needs to be added to the list (17, 11, and 8 are LTS and will remain for a while)
> // // // // // // // //

Reviewer: ```suggestion
To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 19, Java SE 18, Java SE 17, Java SE 11, and Java SE 8, however, the Jakarta EE 10 features are only compatible with Java SE 11 and later.
```
